internal/provider/resources/resource_api_token.go

```diff
@@ -0,0 +1,566 @@
+		resp.State.RemoveResource(ctx)
+		return
+	}
+	if diagnostic != nil {
+		resp.Diagnostics.Append(diagnostic)
+		return
+	}
+
+	diags := data.ReadFromResponse(ctx, apiToken.JSON200)
+	if diags.HasError() {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
+
+	tflog.Trace(ctx, fmt.Sprintf("read an API token resource: %v", data.Id.ValueString()))
+
+	// Save data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r *ApiTokenResource) Update(
+	ctx context.Context,
+	req resource.UpdateRequest,
+	resp *resource.UpdateResponse,
+) {
+	var data models.ApiTokenResource
+
+	// Read Terraform plan data into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Convert Terraform set of roles to API token roles
+	roles, diags := RequestApiTokenRoles(ctx, data.Roles)
+	if diags.HasError() {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
+
+	// Update API token roles
+	updateApiTokenRolesRequest := iam.UpdateApiTokenRolesRequest{
+		Roles: roles,
+	}
+	updatedApiToken, err := r.IamClient.UpdateApiTokenRolesWithResponse(
+		ctx,
+		r.OrganizationId,
+		data.Id.ValueString(),
+		updateApiTokenRolesRequest,
+	)
+	if err != nil {
+		tflog.Error(ctx, "failed to update API token", map[string]interface{}{"error": err})
+		resp.Diagnostics.AddError(
+			"Client Error",
+			fmt.Sprintf("Unable to update API token, got error: %s", err),
+		)
+		return
+	}
+	_, diagnostic := clients.NormalizeAPIError(ctx, updatedApiToken.HTTPResponse, updatedApiToken.Body)
+	if diagnostic != nil {
+		resp.Diagnostics.Append(diagnostic)
+		return
+	}
+
+	// update request
+	updateApiTokenRequest := iam.UpdateApiTokenJSONRequestBody{
+		Name: data.Name.ValueString(),
+	}
+
+	// description
+	if !data.Description.IsNull() {
+		updateApiTokenRequest.Description = data.Description.ValueStringPointer()
+	} else {
+		updateApiTokenRequest.Description = lo.ToPtr("")
+	}
+
+	apiToken, err := r.IamClient.UpdateApiTokenWithResponse(
+		ctx,
+		r.OrganizationId,
+		data.Id.ValueString(),
+		updateApiTokenRequest,
+	)
+	if err != nil {
+		tflog.Error(ctx, "failed to update API token", map[string]interface{}{"error": err})
+		resp.Diagnostics.AddError(
+			"Client Error",
+			fmt.Sprintf("Unable to update API token, got error: %s", err),
+		)
+		return
+	}
+	_, diagnostic = clients.NormalizeAPIError(ctx, apiToken.HTTPResponse, apiToken.Body)
+	if diagnostic != nil {
+		resp.Diagnostics.Append(diagnostic)
+		return
+	}
+
+	// Get api token
+	apiTokenResp, err := r.IamClient.GetApiTokenWithResponse(
+		ctx,
+		r.OrganizationId,
+		data.Id.ValueString(),
+	)
+	if err != nil {
+		tflog.Error(ctx, "failed to update API token", map[string]interface{}{"error": err})
+		resp.Diagnostics.AddError(
+			"Client Error",
+			fmt.Sprintf("Unable to update API token and get API token, got error: %s", err),
+		)
+		return
+	}
+
+	diags = data.ReadFromResponse(ctx, apiTokenResp.JSON200)
```

Comment:
you may need to get the `token` from `req.State` and add it to `apiTokenResp.JSON200`
but i think changing this function to 
```suggestion
	diags = data.ReadFromResponse(ctx, apiTokenResp.JSON200, token)
```
might be better
not sure how good of an idea it is to edit the response of an API - we might mess up somewhere where we don't want to

Reply:
also would need to be done in `Read`

Reply:
do you mean `resp.State`? `req` doesn't have state as a field

also would the values set in create persist in read and update? (im assuming yes because the resource would be managed by tf and it would store all the values associated with it?)

Reply:
nope i mean `req.State` 
its used in line 220

and yes

Reply:
![Screenshot 2024-07-18 at 2 24 40 PM](https://github.com/user-attachments/assets/72ab9e05-d357-4fbd-8824-20d0e595dacc)

Reply:
wouldn't doing `data.Token.ValueString()` also work here? `req.State.Get` populates the prior state data into the target model (which would be the data model we define earlier)

Reply:
sure something like `diags = data.ReadFromResponse(ctx, apiTokenResp.JSON200, data.Token.ValueString())` works

Reply:
in that case we won't need the previous line that edits the response output in create? or do we want both still

Reply:
yea lets just do it this way`diags = data.ReadFromResponse(ctx, apiTokenResp.JSON200, data.Token.ValueString())` for `Create` `Read` `Update`

Reply:
i think create can be `*apiToken.JSON200.Token` since the `data.Token` wouldn't be populated with the new token output. i just pushed my new code so you can see

Reply:
yep yep that looks correct